core/src/main/java/com/linecorp/armeria/internal/common/stream/DefaultHttpDeframer.java

```diff
@@ -0,0 +1,301 @@
+
+    private static final SubscriptionOption[] EMPTY_OPTIONS = {};
+    private static final SubscriptionOption[] POOLED_OBJECTS_OPTIONS = { WITH_POOLED_OBJECTS };
+    private static final SubscriptionOption[] NOTIFY_CANCELLATION_OPTIONS = { NOTIFY_CANCELLATION };
+
+    private final HttpMessageSubscriber subscriber = new HttpMessageSubscriber();
+
+    private final HttpDecoder<T> handler;
+    private final ByteBufDecoderInput input;
+    private final Function<? super HttpData, ? extends ByteBuf> byteBufConverter;
+    private final StreamMessage<? extends HttpObject> publisher;
+
+    @Nullable
+    private RequestHeaders requestHeaders;
+    @Nullable
+    private Subscription upstream;
+
+    private boolean handlerProduced;
+    private boolean sawLeadingHeaders;
+    private boolean initialized;
+    private boolean askedUpstreamForElement;
+    private boolean cancelled;
+
+    /**
+     * Returns a new {@link DefaultHttpDeframer} with the specified {@link HttpDecoder} and
+     * {@link ByteBufAllocator}.
+     */
+    public DefaultHttpDeframer(StreamMessage<? extends HttpObject> streamMessage,
+                               HttpDecoder<T> handler, ByteBufAllocator alloc) {
+        this(streamMessage, handler, alloc, HttpData::byteBuf);
+    }
+
+    /**
+     * Returns a new {@link DefaultHttpDeframer} with the specified {@link HttpDecoder},
+     * {@link ByteBufAllocator} and {@code byteBufConverter}.
+     */
+    public DefaultHttpDeframer(StreamMessage<? extends HttpObject> streamMessage,
+                               HttpDecoder<T> handler, ByteBufAllocator alloc,
+                               Function<? super HttpData, ? extends ByteBuf> byteBufConverter) {
+        publisher = requireNonNull(streamMessage, "streamMessage");
+        this.handler = requireNonNull(handler, "handler");
+        input = new ByteBufDecoderInput(requireNonNull(alloc, "alloc"));
+        this.byteBufConverter = requireNonNull(byteBufConverter, "byteBufConverter");
+        if (publisher instanceof HttpRequest) {
+            requestHeaders = ((HttpRequest) publisher).headers();
+        }
+
+        whenComplete().handle((unused1, cause) -> {
+            if (cause instanceof CancelledSubscriptionException) {
+                cancelAndCleanup();
+            } else {
+                // In addition to 'onComplete()', 'onError()' and 'cancel()',
+                // make sure to call 'cleanup()' even when 'abort()' or 'close()' is invoked directly
+                cleanup();
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public void add(T e) {
+        if (tryWrite(e)) {
+            handlerProduced = true;
+        }
+    }
+
+    @Override
+    protected void subscribe0(EventExecutor executor, boolean withPooledObjects, boolean notifyCancellation) {
+        publisher.subscribe(subscriber, executor, toSubscriptionOptions(withPooledObjects, notifyCancellation));
+        deferredInit();
```

Comment:
We can remove this and rename it to `init()`? because init is always called from `onSubscribe()`?